File: results/2026-04-07_225702/tasks/11-semantic-version-bumper/default-opus/generated-code/semver.py
# Semantic Version Bumper
# Parses version strings, classifies conventional commits, bumps versions,
# and generates changelog entries.

import json
import re
import sys
from dataclasses import dataclass
from enum import Enum
from pathlib import Path


@dataclass
class SemVer:
    """Represents a semantic version (major.minor.patch)."""
    major: int
    minor: int
    patch: int

    def __str__(self):
        return f"{self.major}.{self.minor}.{self.patch}"


class BumpType(Enum):
    """The type of version bump indicated by a commit."""
    MAJOR = "major"
    MINOR = "minor"
    PATCH = "patch"


@dataclass
class Commit:
    """A parsed conventional commit."""
    bump_type: BumpType
    type: str | None      # e.g. "feat", "fix", "chore"
    description: str
    scope: str | None


# Maps conventional commit types to bump levels
_BUMP_MAP = {"feat": BumpType.MINOR}
# Everything else (fix, chore, docs, refactor, etc.) defaults to PATCH


def parse_commit(message: str) -> Commit:
    """Parse a conventional commit message and determine the bump type.

    Supports: type[(scope)][!]: description
    A trailing '!' or 'BREAKING CHANGE:' footer triggers a major bump.
    """
    first_line = message.split("\n")[0]
    is_breaking = "BREAKING CHANGE:" in message or "BREAKING-CHANGE:" in message

    # Match conventional commit pattern: type(scope)!: description
    match = re.match(r"^(\w+)(?:\(([^)]+)\))?(!)?\s*:\s*(.+)$", first_line)
    if not match:
        # Non-conventional commit — treat as patch
        return Commit(BumpType.PATCH, None, first_line.strip(), None)

    ctype, scope, bang, desc = match.groups()
    if bang or is_breaking:
        bump = BumpType.MAJOR
    else:
        bump = _BUMP_MAP.get(ctype, BumpType.PATCH)

    return Commit(bump, ctype, desc.strip(), scope)


def bump_version(current: SemVer, commit_messages: list[str]) -> SemVer:
    """Determine the next version by analyzing commit messages.

    The highest bump type wins: major > minor > patch.
    Major resets minor+patch, minor resets patch.
    """
    if not commit_messages:
        raise ValueError("No commits provided — cannot determine version bump")

    # Parse all commits and find the highest bump level
    # MAJOR > MINOR > PATCH, so we use ordering: MAJOR=0, MINOR=1, PATCH=2
    priority = {BumpType.MAJOR: 0, BumpType.MINOR: 1, BumpType.PATCH: 2}
    parsed = [parse_commit(msg) for msg in commit_messages]
    highest = min(parsed, key=lambda c: priority[c.bump_type]).bump_type

    if highest == BumpType.MAJOR:
        return SemVer(current.major + 1, 0, 0)
    elif highest == BumpType.MINOR:
        return SemVer(current.major, current.minor + 1, 0)
    else:
        return SemVer(current.major, current.minor, current.patch + 1)


def parse_version(version_str: str) -> SemVer:
    """Parse a semantic version string like '1.2.3' or 'v1.2.3'."""
    match = re.match(r"^v?(\d+)\.(\d+)\.(\d+)$", version_str.strip())
    if not match:
        raise ValueError(f"Invalid semantic version: '{version_str}'")
    return SemVer(int(match.group(1)), int(match.group(2)), int(match.group(3)))


def read_version_file(path: str) -> SemVer:
    """Read a version from a VERSION file or package.json."""
    p = Path(path)
    if not p.exists():
        raise FileNotFoundError(f"Version file not found: {path}")

    content = p.read_text()

    if p.name == "package.json":
        data = json.loads(content)
        if "version" not in data:
            raise ValueError(f"No 'version' field in {path}")
        return parse_version(data["version"])

    # Plain text version file (e.g. VERSION)
    return parse_version(content.strip())


def write_version_file(path: str, version: SemVer) -> None:
    """Write the new version to a VERSION file or package.json."""
    p = Path(path)

    if p.name == "package.json":
        data = json.loads(p.read_text())
        data["version"] = str(version)
        p.write_text(json.dumps(data, indent=2) + "\n")
    else:
        p.write_text(str(version) + "\n")


# Changelog section headers mapped from commit types
_SECTION_MAP = {
    "feat": "Features",
    "fix": "Bug Fixes",
}


def generate_changelog(version: SemVer, commit_messages: list[str]) -> str:
    """Generate a markdown changelog entry grouped by commit type."""
    if not commit_messages:
        raise ValueError("No commits provided — cannot generate changelog")

    parsed = [parse_commit(msg) for msg in commit_messages]

    # Group commits by section
    sections: dict[str, list[Commit]] = {}
    for c in parsed:
        section = _SECTION_MAP.get(c.type, "Other") if c.type else "Other"
        sections.setdefault(section, []).append(c)

    lines = [f"## {version}\n"]
    # Render known sections first in a stable order, then "Other"
    for section in ["Features", "Bug Fixes", "Other"]:
        if section not in sections:
            continue
        lines.append(f"### {section}\n")
        for c in sections[section]:
            prefix = "**BREAKING** " if c.bump_type == BumpType.MAJOR else ""
            scope = f"**{c.scope}**: " if c.scope else ""
            lines.append(f"- {prefix}{scope}{c.description}")
        lines.append("")

    return "\n".join(lines).rstrip() + "\n"


def load_commits(path: str) -> list[str]:
    """Load commit messages from a file (one per line, blank lines skipped)."""
    p = Path(path)
    if not p.exists():
        raise FileNotFoundError(f"Commits file not found: {path}")
    return [line for line in p.read_text().splitlines() if line.strip()]


def main():
    """CLI entry point: semver.py <version-file> <commits-file>"""
    if len(sys.argv) != 3:
        print("Usage: semver.py <version-file> <commits-file>", file=sys.stderr)
        sys.exit(1)

    version_file, commits_file = sys.argv[1], sys.argv[2]

    try:
        current = read_version_file(version_file)
        commits = load_commits(commits_file)
        new_version = bump_version(current, commits)
        write_version_file(version_file, new_version)
        changelog = generate_changelog(new_version, commits)

        print(f"{current} -> {new_version}\n")
        print(changelog)
    except (ValueError, FileNotFoundError) as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: results/2026-04-08_114024/tasks/11-semantic-version-bumper/default-opus/generated-code/semver.ts
// Semantic version bumper — core library.
// Parses conventional commits, bumps semver, updates version files,
// and generates changelog entries.

// ── Types ────────────────────────────────────────────────────────────────────

export interface SemVer {
  major: number;
  minor: number;
  patch: number;
}

// ── Version parsing ──────────────────────────────────────────────────────────

const SEMVER_RE = /^v?(\d+)\.(\d+)\.(\d+)$/;

/** Parse a "major.minor.patch" string (optional leading "v"). */
export function parseVersion(raw: string): SemVer {
  const m = SEMVER_RE.exec(raw.trim());
  if (!m) throw new Error(`Invalid semantic version: "${raw}"`);
  return { major: Number(m[1]), minor: Number(m[2]), patch: Number(m[3]) };
}

/** Format a SemVer object back to a string. */
export function formatVersion(v: SemVer): string {
  return `${v.major}.${v.minor}.${v.patch}`;
}

// ── Commit parsing ───────────────────────────────────────────────────────────

export interface Commit {
  hash: string;
  message: string;
}

export interface ClassifiedCommits {
  major: Commit[];
  minor: Commit[];
  patch: Commit[];
}

export type BumpType = "major" | "minor" | "patch";

/**
 * Parse a git log (one commit per line: "<hash> <message>") into Commit objects.
 * Multi-line messages (e.g. with BREAKING CHANGE footers) are supported —
 * continuation lines that don't start with a hash are appended to the
 * previous commit's message.
 */
export function parseCommitLog(raw: string): Commit[] {
  const commits: Commit[] = [];
  // Matches lines starting with a hex hash (7+ chars) followed by a space
  const COMMIT_LINE = /^([0-9a-f]{7,}) (.+)$/;

  for (const line of raw.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) continue;

    const m = COMMIT_LINE.exec(trimmed);
    if (m) {
      commits.push({ hash: m[1], message: m[2] });
    } else if (commits.length > 0) {
      // Continuation line (e.g. BREAKING CHANGE footer) — append to previous
      commits[commits.length - 1].message += "\n" + trimmed;
    }
  }
  return commits;
}

// Conventional commit prefix regex:  type(scope)?: message  or  type(scope)!: message
const CONVENTIONAL_RE = /^(\w+)(?:\([^)]*\))?(!)?: /;

/**
 * Classify commits by their impact level according to conventional commits:
 *   - "!" suffix or BREAKING CHANGE footer → major
 *   - "feat" type → minor
 *   - "fix" type → patch
 *   - everything else is ignored (docs, chore, ci, etc.)
 */
export function classifyCommits(commits: Commit[]): ClassifiedCommits {
  const result: ClassifiedCommits = { major: [], minor: [], patch: [] };

  for (const commit of commits) {
    // Check for BREAKING CHANGE footer anywhere in the message
    if (commit.message.includes("BREAKING CHANGE")) {
      result.major.push(commit);
      continue;
    }

    const m = CONVENTIONAL_RE.exec(commit.message);
    if (!m) continue; // non-conventional commit, skip

    const type = m[1];
    const bang = m[2]; // "!" if present

    if (bang) {
      result.major.push(commit);
    } else if (type === "feat") {
      result.minor.push(commit);
    } else if (type === "fix") {
      result.patch.push(commit);
    }
    // other types (docs, chore, etc.) are intentionally ignored
  }

  return result;
}

// ── Version bumping ──────────────────────────────────────────────────────────

/** Determine the highest-priority bump needed, or null if no bump is needed. */
export function determineBump(classified: ClassifiedCommits): BumpType | null {
  if (classified.major.length > 0) return "major";
  if (classified.minor.length > 0) return "minor";
  if (classified.patch.length > 0) return "patch";
  return null;
}

/** Apply a bump to a version, following semver rules (reset lower fields). */
export function bumpVersion(v: SemVer, bump: BumpType): SemVer {
  switch (bump) {
    case "major":
      return { major: v.major + 1, minor: 0, patch: 0 };
    case "minor":
      return { major: v.major, minor: v.minor + 1, patch: 0 };
    case "patch":
      return { major: v.major, minor: v.minor, patch: v.patch + 1 };
  }
}

// ── Version file I/O ─────────────────────────────────────────────────────────

/** Read the current version from a file (plain text or package.json). */
export async function readVersionFile(path: string): Promise<SemVer> {
  const content = await Bun.file(path).text();

  if (path.endsWith(".json")) {
    const pkg = JSON.parse(content);
    if (!pkg.version) throw new Error(`No "version" field found in ${path}`);
    return parseVersion(pkg.version);
  }

  // Plain text file — expect just the version string
  return parseVersion(content.trim());
}

/** Write a version to a file (plain text or package.json, preserving fields). */
export async function writeVersionFile(
  path: string,
  version: SemVer
): Promise<void> {
  const vStr = formatVersion(version);

  if (path.endsWith(".json")) {
    const content = await Bun.file(path).text();
    const pkg = JSON.parse(content);
    pkg.version = vStr;
    await Bun.write(path, JSON.stringify(pkg, null, 2) + "\n");
  } else {
    await Bun.write(path, vStr + "\n");
  }
}

// ── Changelog generation ─────────────────────────────────────────────────────

/** Strip the conventional commit prefix to get just the human-readable part. */
function stripPrefix(msg: string): string {
  // Take only the first line (ignore footers)
  const firstLine = msg.split("\n")[0];
  return firstLine.replace(CONVENTIONAL_RE, "");
}

/** Generate a markdown changelog entry for a release. */
export function generateChangelog(
  version: string,
  classified: ClassifiedCommits
): string {
  const lines: string[] = [];
  const date = new Date().toISOString().slice(0, 10);

  lines.push(`## ${version} (${date})`);
  lines.push("");

  if (classified.major.length > 0) {
    lines.push("### Breaking Changes");
    for (const c of classified.major) lines.push(`- ${stripPrefix(c.message)} (${c.hash})`);
    lines.push("");
  }

  if (classified.minor.length > 0) {
    lines.push("### Features");
    for (const c of classified.minor) lines.push(`- ${stripPrefix(c.message)} (${c.hash})`);
    lines.push("");
  }

  if (classified.patch.length > 0) {
    lines.push("### Bug Fixes");
    for (const c of classified.patch) lines.push(`- ${stripPrefix(c.message)} (${c.hash})`);
    lines.push("");
  }

  return lines.join("\n");
}
